daml-lf/spec/contract-id.rst

```diff
@@ -0,0 +1,245 @@
+construction. It assumes a hash-based MAC function HMAC, and defines
+for each big-endian 32 bits integer ``n`` the derived seed::
+
+  deriveSeed(seed, n) := HMAC(seed, n)
+  
+
+Submission seed
+---------------
+
+Any random number with sufficient entropy works as submission seed
+``submission_seed`` (256 bit entropy).
+
+Transaction seed construction
+-----------------------------
+
+From the submission, a so-called *transaction seed* is derived as follows::
+
+  deriveTransactionSeed(submissionSeed, participandId, submissionTime, submitters) :=
+     HMAC(submissionSeed, participantId ∥ submissionTime ∥ nₛ ∥ submitters)
+
+where
+
+* ``submissionSeed`` is the submission seed;
+* ``participandId`` is the participant ID;
+* ``submissionTime`` is the submission time in micro second encoded as
+  a 64 bytes big-endian integer;
+* ``nₛ`` is the number of submitters encoded as a 32 big-endian bits
+  integer;
+* ``submitters`` is the concatenation of the submitter IDs sorted
+  lexicographically.
+
+Strings such as participant IDs or submitter IDs are interpreted as
+their US-ASCII encoding prefixed with there size encoded as a
+32 bits big-endian integer.
+  
+     
+Derivation of seeds for root nodes the transaction
+--------------------------------------------------
+
+For each root node of the transaction, a root seed `rootSeedᵢ` is
+computed from the initial seed as follows::
+
+  rootSeedᵢ = deriveSeed(transactionSeed, i)
+
+Derivation of seeds for the children of exercise nodes
+------------------------------------------------------
+
+For an exercise node with seed ``nodeSeed``, the seeds `childSeedᵢ` for the
+children are derived as follows::
+
+  childSeedᵢ = deriveSeed(nodeSeed, i)
+
+Derivation of contract ID discriminator
+---------------------------------------
+
+The discriminator for the created contract ID is given as follows::
+
+  discriminator = HMAC(nodeSeed, submissionTime ∥ nₛ ∥ stackholders) 
+
+where
+
+* ``nodeSeed`` is the seed of the node where the contract is created;
+* ``submissionTime`` is the submission time in micro second encoded as
+  a 64 bytes big-endian integer;
+* ``nₛ`` is the number of stakeholder's of the contract encoded as a
+  32 bits big-endian integer;
+* ``stackholders`` is the concatenation of the stakeholders IDs sorted
+  lexicographically.
+
+IDs of stakeholder are interpreted as their US-ASCII encoding prefixed
+with there size encoded as a 32 bits  big-endian integer.
+  
+Submission
+^^^^^^^^^^
+
+The submission the following steps:
+
+* Peek a submission seed  with high entropy.
+* Derive the transaction seed and start the interperation
+* During interpretation derive the node seeds, and the discriminator
+  of local contract ID accordingly the scheme described above.  If the
+  discriminator is not `fresh <Discriminator Freshness_>`_, abort the
+  interpretation. The submitter can restart the interpretation with
+  another submission seed.
```

Comment:
According to step 1, the submission seed is picked by "submission". This sounds as if the submitter does not have control over the seed and therefore. This reads as if the submitter could choose the submission seed. So I recommend to either change the first step into `Submission takes a high-entropy submission seed as input` or to change this into 
```suggestion
  interpretation. The submitter can restart the interpretation, which will pick
  another submission seed.
```